use disk at query time of notification payload creation

dpsutton opened the pull request:
### OOMs from Notifications

A short history of how we got here:
We render alerts and dashboards with a graal js interpreter. To do this we need the resultsets of queries. We used to just gather these all in memory and then render them. This works super well for small datasets and exposes the instance to OOM issues if a large query is returned.

Ngoc (possibly others as well) made a nice addition in https://github.com/metabase/metabase/pull/51708 that uses disk to ensure that only a single resultset is in memory at any given time. the gist is run query, if results greater than some row count threshold, dump to disk, and then during rendering refetch it into memory.

This still leaves us susceptible to OOM if any particular resultset is too large.

What ends up happening is that instances randomly die and emails are not sent.

### The fix here

The gist of this fix is quite simple: if resultsets start getting to big, stream them to disk as they come from the db rather than after they are done. Then add two knobs: a) the size limit that we don't attempt to render them and b) whether to enforce this limit. This lets anyone easily go back to the previous behavior if they want to allow large query results.

#### How it is implemented

We run our queries in a transducing context here: https://github.com/metabase/metabase/blob/master/src/metabase/query_processor/pipeline.clj#L78-L87 and the reducing function is created just above with `(rff metadata)` which returns a reducing function.

So this PR creates a new one which has two modes: collecting results in memory, similar to the default one (https://github.com/metabase/metabase/blob/master/src/metabase/query_processor/reducible.clj#L15-L38) which is optionally wired up here: https://github.com/metabase/metabase/blob/master/src/metabase/query_processor.clj#L81

If the number of rows remains under 2000, this reducing function behaves identically to the default one. If more rows are returned, it takes all accumulated rows so far and starts serializing them to disk. In this way we have prevent having entire resultsets in memory and we can also inspect the entire size and decide if we should abandon trying to render the results.

There is a limit at which we won't open the files (by default 10mb). And the reducing function will monitor the filesize at is creates it and stop running the query when the file is 1.3x this limit. This helps us not fill up disk with query results we won't attempt to render anyways.


## What it looks like

This dashboard has a query that generates 50mb of data and then a simple line chart of orders aggregated by month and averaged.

<img width="1382" height="1008" alt="image" src="https://github.com/user-attachments/assets/4155cafc-cde1-4c7c-8d5b-844a8c4e3729" />

The email comes across as

<img width="1382" height="1008" alt="image" src="https://github.com/user-attachments/assets/ef64db7d-3385-4a11-8c70-dece9dedc26a" />


The logs output are

```
2025-10-07 16:45:28,775 INFO notification.send :: Sending {mb-notification_id=, mb-payload_type=:notification/dashboard}
2025-10-07 16:45:29,795 INFO payload.temp-storage :: Row count reached threshold (2000), switching to streaming mode {mb-card_id=311, mb-dashboard_id=54, mb-notification_id=, mb-payload_type=:notification/dashboard}
2025-10-07 16:45:30,102 WARN payload.temp-storage :: Results have exceeded 1.3 times of `notification-temp-file-size-max-bytes` of 13.4 mb (max: 10.0 mb). Truncating query results. ({:dashboard_id 54, :card_id 311, :dashcard_id 145}) {mb-card_id=311, mb-dashboard_id=54, mb-notification_id=, mb-payload_type=:notification/dashboard}
2025-10-07 16:45:30,147 INFO payload.temp-storage :: 💾 Stored 40000 rows to disk: 13.42 MB (never loaded into memory) (note query results were truncated) {mb-card_id=311, mb-dashboard_id=54, mb-notification_id=, mb-payload_type=:notification/dashboard}
2025-10-07 16:45:30,358 INFO payload.temp-storage :: ✓ Completed with 12 rows in memory (under threshold) {mb-card_id=355, mb-dashboard_id=54, mb-notification_id=, mb-payload_type=:notification/dashboard}
2025-10-07 16:45:31,141 INFO payload.temp-storage :: Row count reached threshold (2000), switching to streaming mode {mb-card_id=356, mb-dashboard_id=54, mb-notification_id=, mb-payload_type=:notification/dashboard}
2025-10-07 16:45:31,237 INFO payload.temp-storage :: 💾 Stored 18760 rows to disk: 7.68 MB (never loaded into memory) {mb-card_id=356, mb-dashboard_id=54, mb-notification_id=, mb-payload_type=:notification/dashboard}
2025-10-07 16:45:31,253 WARN payload.temp-storage :: ⚠️  SKIPPING LOAD - File too large: 13.42 MB (max: 10.00 MB). File will NOT be loaded into memory. {mb-channel_type=:channel/email, mb-handler_id=, mb-notification_id=, mb-payload_type=:notification/dashboard}
2025-10-07 16:45:31,253 WARN channel.shared :: 🚫 Result file too large (13.42 MB > 10.00 MB max). Skipping load to protect memory. {mb-channel_type=:channel/email, mb-handler_id=, mb-notification_id=, mb-payload_type=:notification/dashboard}
2025-10-07 16:45:31,255 ERROR render.card :: Pulse card query error: results too large {mb-card_id=311, mb-channel_type=:channel/email, mb-handler_id=, mb-notification_id=, mb-payload_type=:notification/dashboard}
2025-10-07 16:45:33,572 INFO payload.temp-storage :: 📂 Loading streamed results from disk: 7.68 MB {mb-channel_type=:channel/email, mb-handler_id=, mb-notification_id=, mb-payload_type=:notification/dashboard}
2025-10-07 16:45:33,637 INFO payload.temp-storage :: ✅ Loaded 18760 rows from disk in 65 ms (7.68 MB) {mb-channel_type=:channel/email, mb-handler_id=, mb-notification_id=, mb-payload_type=:notification/dashboard}
2025-10-07 16:45:33,995 INFO notification.send :: Sent successfully {mb-channel_type=:channel/email, mb-handler_id=, mb-notification_id=, mb-payload_type=:notification/dashboard}
2025-10-07 16:45:34,005 INFO notification.send :: Done processing notification {mb-notification_id=, mb-payload_type=:notification/dashboard}
2025-10-07 16:45:34,014 DEBUG middleware.log :: POST /api/pulse/test 200 5314ms (159 DB calls) App DB connections: 1/6 Jetty threads: 5/50 (2 idle, 0 queued) (176 total active threads) Queries in flight: 0 (0 queued); h2 DB 1 connections: 0/1 (0 threads blocked) {:metabase-user-id 1}
```

The relevant logs for these two cards are:

#### the 50mb query
`Row count reached threshold (2000), switching to streaming mode` tried to not use disk but we should switch
`Results have exceeded 1.3 times of `notification-temp-file-size-max-bytes` of 13.0 mb (max: 10.0 mb). Truncating query results.` we have exceed the size that we would read off of disk, so stop running the query
`💾 Stored 400338 rows to disk: 13.00 MB (never loaded into memory) (note query results were truncated) {mb-card_id=311, mb-dashboard_id=54, mb-notification_id=, mb-payload_type=:notification/dashboard}` a summary of what is stored to disk, including notification id, dashboard, etc. This one does not have a notification id because i'm running the preview "send now" version.

#### The small query
`✓ Completed with 12 rows in memory (under threshold) {mb-card_id=355, mb-dashboard_id=54, mb-notification_id=, mb-payload_type=:notification/dashboard}` Really nothing to see here. Simple line graph, 12 query results, no need to involve disk, etc. Happy path is very happy

#### Rendering
We have a few lines about the first card's results being too large and being skipped
```
⚠️  SKIPPING LOAD - File too large: 13.00 MB (max: 10.00 MB). File will NOT be loaded into memory. {mb-channel_type=:channel/email, mb-handler_id=, mb-notification_id=, mb-payload_type=:notification/dashboard}
🚫 Result file too large (13.00 MB > 10.00 MB max). Skipping load to protect memory. {mb-channel_type=:channel/email, mb-handler_id=, mb-notification_id=, mb-payload_type=:notification/dashboard}
Pulse card query error: results too large {mb-card_id=311, mb-channel_type=:channel/email, mb-handler_id=, mb-notification_id=, mb-payload_type=:notification/dashboard}
```


Optimizations for the future:
- explore how parallel we can get now. Presumably we are now quite memory effecient. Can we go back to 3, or 10, worker threads?

## Correctness
Query results are coming back identical as long as they aren't truncated. Of course once we truncate they should not be the same. And we get a nice speed up as well. A query that returns 50mb of data:

```clojure
(time (->
        (metabase.notification.payload.execute/execute-card 1 311)
        :result
        (select-keys [:notification/truncated?
                      :status :running_time
                      :data.rows-file-size
                      :data])
        (update :data (fn [d] (select-keys d [:rows])))))
"Elapsed time: 959.727584 msecs"
{:notification/truncated? true,
 :status :completed,
 :running_time 902,
 :data.rows-file-size 14073732,
 :data {:rows #object[metabase.notification.payload.temp_storage.StreamingTempFileStorage
                      "0x23bdac12"
                      {:status :pending, :val nil}]}}
temp-storage=> (-> *1 :data :rows str)
"#StreamingTempFileStorage{:file notification-10650480272503863452.npy, :size 13743.88 KB, :context {:card-id 311}}"
temp-storage=> (time
                 (-> (toucan2.core/select-one-fn :dataset_query :model/Card :id 311)
                     (metabase.query-processor/process-query)
                     :data :rows count))
"Elapsed time: 4272.368625 msecs"
500000
```

Here it takes us 959 ms to abandon a query after 10mb. But the regular query processor spends 4272ms processing a query that would create 50mb of data.

## Performance
At first i was using gzip to keep results smaller. This has two issues: performance and correctness. For performance, it was taking a 4 second query to 18 seconds and murdering performance. For correctness, we do want to have a good sense of how big results are in a csv or in memory. GZip is really good at compressing. So a few results that would be 50mb in a csv were ~8mb on disk gzipped. Which means it wasn't a great proxy for if the query might OOM the instance, and it also slowed everything down.

## Size:

The following query is my "50mb" query:

```sql
SELECT 
     i AS id,
     'User_' || i AS username,
     'user' || i || '@example.com' AS email,
     'This is a long description for row ' || i || '. ' ||
     'It contains enough text to make the row size larger. ' ||
     'We want to generate approximately 50MB of data to test ' ||
     'the file size limit protection in the streaming system. ' ||
     'This query uses generate_series to create many rows.' AS description,
     CASE WHEN i % 3 = 0 THEN 'Active'
          WHEN i % 3 = 1 THEN 'Pending'
          ELSE 'Inactive' END AS status,
     (RANDOM() * 1000)::numeric(10,2) AS amount,
     NOW() - (i || ' days')::interval AS created_at
   FROM generate_series(1, 500000) AS i;
```

When i download this as a csv it takes 165M, when saved as a nippy file it takes:

```clojure
(time (->
        (metabase.notification.payload.execute/execute-card 1 311)
        :result
        (select-keys [:notification/truncated?
                      :status :running_time
                      :data.rows-file-size
                      :data])
        (update :data (fn [d] (select-keys d [:rows])))))
"Elapsed time: 6685.622292 msecs"
{:notification/truncated? true,
 :status :completed,
 :running_time 6616,
 :data.rows-file-size 136600169,
 :data {:rows #object[metabase.notification.payload.temp_storage.StreamingTempFileStorage
                      "0x7367674f"
                      {:status :pending, :val nil}]}}
temp-storage=> (-> *1 :data :rows (.file) (.length) (human-readable-size))
"130.3 mb"
```

For card 356 which is people join orders join products i see

```clojure
temp-storage=> 
(time (->
        (metabase.notification.payload.execute/execute-card 1 356)
        :result
        (select-keys [:notification/truncated?
                      :status :running_time
                      :data.rows-file-size
                      :data])
        (update :data (fn [d] (select-keys d [:rows])))))
"Elapsed time: 1152.316792 msecs"
{:status :completed,
 :running_time 993,
 :data.rows-file-size 8056845,
 :data {:rows #object[metabase.notification.payload.temp_storage.StreamingTempFileStorage
                      "0x34b0eb54"
                      {:status :pending, :val nil}]}}
temp-storage=> (-> *1 :data :rows (.file) (.length) (human-readable-size))
"7.7 mb"
```
and the download sizes are:

|format|size|
|----|----|
|csv|7.2M|
|xlsx|4.1M|
|json|19M|

So roughly in-line with other formats.

## Weird types:
How does this work with pgarrays, weird bigquery types, etc.? I suspect it has the same issues that caching does, which is to say it blows up, but also it will be limited by what can be fed to the js interpreter. That turns everything into json so we probably just need to verify 

- nippy serialization
- cheshire serialization

If these two work, then the pipeline works.

```sql
SELECT 
     i AS id,
     'User_' || i AS username,
     -- JSON types
     ('{"userId": ' || i || ', "score": ' || (i % 1000) || ', "active": ' || (i % 2 = 0)::text || '}')::jsonb AS settings,
     -- Arrays
     ARRAY['tag' || (i % 10), 'category' || (i % 5)]::text[] AS tags,
     ARRAY[i, i*2, i*3]::integer[] AS numbers,
     -- UUID
     gen_random_uuid() AS uuid,
     -- Network
     ('192.168.' || (i % 255) || '.' || ((i*7) % 255))::inet AS ip,
     -- Geometric
     point(i % 180 - 90, i % 360 - 180) AS coordinates,
     circle(point(i % 100, i % 100), 50) AS area,
     -- Text search
     to_tsvector('english', 'Content for row ' || i) AS search_data,
     -- Binary
     decode(md5(i::text), 'hex') AS hash,
     -- Range types
     int4range(i, i + 100) AS value_range,
     -- Money
     ((RANDOM() * 1000)::numeric(10,2))::money AS price,
     -- Standard
     NOW() - ((i % 365) || ' days')::interval AS created_at
FROM generate_series(1, 5000) AS i;
```

```clojure
(time (->
        (metabase.notification.payload.execute/execute-card 1 357)
        :result
        (select-keys [:notification/truncated?
                      :status :running_time
                      :data.rows-file-size
                      :data])
        (update :data (fn [d] (select-keys d [:rows])))
        :data :rows deref first))
"Elapsed time: 372.73275 msecs"
[1
 "User_1"
 "{\"score\": 1, \"active\": false, \"userId\": 1}"
 ["tag1" "category1"]
 [1 2 3]
 #uuid "8d4b0e23-9f99-4d78-a059-fd2187b22e61"
 "192.168.1.7"
 "(-89.0,-179.0)"
 "<(1.0,1.0),50.0>"
 "'1':4 'content':1 'row':3"
 #object["[B" "0x75d67e3a" "[B@75d67e3a"]
 "[1,101)"
 827.08M
 "2025-10-06T16:54:35.640219Z"]
```

dpsutton committed:
use disk at query time of notification payload creation

We have some nice work to dump query results to disk if they are larger
than a row-count threshold. But this presupposes that results can fit
into memory. It's completely possible, and relatively easy, to do some
kind of "select *" operation in a dashboard where the only limit that
prevents billions of rows coming back is the excel limit we add in.

This tries to optimistically use the disk. If creates a new rff that has
two modes: in-memory and a gzipped dataoutput stream. The first is in
operation for the first 2,000 rows. If we never exceed this limit it
remains a regular reducing function. Once this threshold is exceeded, we
open a new file and stuff the rows in there and don't keep them in
memory any longer. This prevents us from having to hold the entire
result set in memory at once.

On the consumption side, we have a threshold such that if a file is
_larger_ than this, we won't attempt to load it and render it for static
viz. This is what lets us defensively decline to do an operation that
would reasonably lead to an OOM.

Optimizations for the future:
- give up on file creation once it exceeds the threshold size. We can
easily monitor the size of the file as we go and stop the transduction
or results.
- clear out files more frequently. This will probably litter lots of
temp files, some possibly quite large, on teh disk. We need to more
frequently clear them up. The directory is marked as deleteOnExit, but
we need to ensure we don't put lots of files in there during process
lifetime.
- better error rendering of "too large" error. Right now it hits the
default error rendering (previously it would oom). We can make this
better.
- explore how parallel we can get now. Presumably we are now quite
memory effecient. Can we go back to 3, or 10, worker threads?

dpsutton committed:
fix typo and reintroduce constant for row threshold

dpsutton committed:
more references to typo constant

dpsutton committed:
missed one somehow?

dpsutton committed:
include pulses (alerts) for autocommit off for downloads

we want to stream results. and now we do into a file so this can respect
the page size which only happens in pg when using autocommit false

dpsutton committed:
special error msg for too large results/dont attach large results

feedback that query results were too large

alerts also by default include results. let's check for the
render/too-large and not include them in that case. Was actually getting
a seq error on the deftype that holds the file. maybe want to give that
a seqable method?

dpsutton committed:
fix tests

we had an unrealized temp storage file going through

dpsutton committed:
unification all to streaming before moving to temp-storage

dpsutton committed:
move streaming to the temp-storage namespace

dpsutton committed:
don't deref results if no one wants attachments

dpsutton committed:
don't log the error here. it's pretty normal

dpsutton committed:
doc strings

dpsutton committed:
better logging and truncate results above threshold

once we are above the size that we won't read the file, we don't have to
keep piping results to it.

Concretely, what does this get us? If someone absentmindedly did
`select * from transactions` it could be 100s of GBs or more. Without
this, we would fill up the disk for the instance even though we don't
try to render it in the email. But once we know that there's no way we
will read the contents of the file, we really don't have to write to it
any longer.

dpsutton committed:
Make settings for size limit and whether to enforce

maybe someone has a beefy machine and they don't mind chugging through
843mb of query results to get a graph. they can easily set
`MB_ENFORCE_NOTIFICATION_TEMP_FILE_SIZE_LIMIT` to false and the old
behavior will be restored. If someone doesn't like the 10mb limit and
prefers a 50mb limit they can do
`MB_NOTIFICATION_TEMP_FILE_SIZE_MAX_BYTES=52428800` and then they will
have the higher limit

dpsutton committed:
Huge performance improvements

The real gist here is removing the gzip on the output. This was crushing
our performance. 18 seconds instead of ~4 seconds. Real bad. And it's
better to not gzip anyways. We really want to be tied pretty closely to
the size of the results. GZiping uses a ton of cpu and gets it very
compact. But we want a number that is pretty representative of the
results. Repeated values get gzipped down to almost nothing and then
they could still OOM the instance. But 50mb of results might get
squashed down to ~8mb of text. We want to be tethered close the size in
a csv or in memory.

Also changed the serialization to write chunks of rows and not a single
row. Its totally possible that the gzip fix did all of the work and
chunks or individual rows all get put together into the
bufferedoutputstream so it doesn't matter anyways. But the good thing is
that the format of this file is entirely the work of this namespace and
no one else gets to see it.

dpsutton committed:
Check for fidelity of regular results and disk based

timing wise: running the same card 356 (a join of people, products,
orders) and verify timing stays consistent (we are only 100ms longer
with disk)

```clojure
temp-storage=> (time
                 (-> (toucan2.core/select-one-fn :dataset_query :model/Card :id 356)
                     metabase.query-processor/process-query
                     :data :rows count))
"Elapsed time: 740.709583 msecs"
18760
temp-storage=>
(time (->
        (metabase.notification.payload.execute/execute-card 1 356)
        :result
        (select-keys [:notification/truncated?
                      :status :running_time
                      :data.rows-file-size
                      :data])
        (update :data (fn [d] (select-keys d [:rows])))))
"Elapsed time: 840.608 msecs"
{:status :completed,
 :running_time 724,
 :data.rows-file-size 8056845,
 :data {:rows #object[metabase.notification.payload.temp_storage.StreamingTempFileStorage
                      "0x3336077d"
                      {:status :pending, :val nil}]}}
```

Compare in memory results:

```clojure
temp-storage=> (let [card-id 356
                     regular-rows (-> (toucan2.core/select-one-fn :dataset_query :model/Card :id card-id)
                                      (metabase.query-processor/process-query)
                                      :data :rows)
                     disk-rows    (-> (metabase.notification.payload.execute/execute-card 1 356)
                                      :result :data :rows)
                     dereffed @disk-rows]
                 (doseq [[f results] [["/tmp/results.from-disk" dereffed]
                                      ["/tmp/results.from-qp" regular-rows]]]
                   (with-open [w (clojure.java.io/writer f)]
                     (doseq [row results]
                       (io/copy (str (pr-str row) \newline) w))))
                 [(= regular-rows dereffed) (= (count regular-rows) (count dereffed))])
[true true]
```

and then compare the files written above:

```shell
/tmp
❯ diff results.from-disk results.from-qp

/tmp
❯ echo $?
0
```

dpsutton committed:
weird postgres types

dpsutton committed:
postgres 12 didn't like gen_random_uuid so just hardcode one

everyone else promised they wouldn't use
7e3cd49d-bfe1-4620-83dd-0c163719175c for anything important

dpsutton committed:
formatting, lints

dpsutton committed:
alignment

dpsutton committed:
reformat, remove unused aliases

dpsutton committed:
remove errant comment

sansolovyov reviewed (approved):
At first RFF tripped me up somewhat being quite big, but once you start reading, the logic is sound and the code is clean. I hope I'm not missing anything important, and all I have is little nitpicks, this is really great!

piranha commented on src/metabase/notification/payload/temp_storage.clj:
```diff
@@ -68,32 +142,166 @@
+            ;; Already streaming - write row directly to file
+
+            (let [{:keys [^DataOutputStream output-stream ^File file]} @streaming-state]
+              (vswap! rows conj! row)
```
Also if you do not mind, I'd prefer `(if-not (zero? ....` in the next line, this way there is an early return of `result` and it'll be easier to visually match what's going on.

dpsutton committed:
cleanup logic and add some comments

unconditionally add to rowcount and in-memory collection of rows. For
two of three branches, this is "done": (1) collecting in memory, and (2)
streaming to disk but our "batch" isn't full. Only when the batch is a
multiple of 5,000 rows (number completely picked at random, balancing
this) we can then write the row block to disk.

Totally possible this can switch to just writing to disk on each row and
rely on the buffered writer to handle batching. But the use of the
in-memory collection being similar across almost all of the branches
makes this solution feel a bit cleaner for the moment.

also does things like inverting `if-not` so the short branch is at the
top and not hanging at the end where it's esay to miss.

dpsutton committed:
unify two file size settings

used to have some notion of `enforce` and `file-size-limits` Rather than
juggle both, you can just set the filesize to 0 to make it unbounded

dpsutton committed:
reuse human-readable-size in more places, fix some typos

dpsutton committed:
typos, fixes

dpsutton committed:
prometheus metrics

dpsutton committed:
followup to metrics

I’ve added a new metric
```clojure
   (prometheus/counter :metabase-notification/temp-storage
                       {:description "Number and type of temporary storage uses"
                        ;; memory, disk, above-threshold, truncated, not-limited
                        :labels [:storage]})
```

to record the different scenarios of query results:
- stayed in `memory`
- wrote them to `disk` but remained under the threshold
- wrote to disk and they went `above-threshold`
- wrote to disk and they went so far above threshold that we abandoned the query so the results were `truncated`
- wrote to disk but the limit was set to 0 so the results were `not-limited`

and these look like: (replaced # with ♯ so git message will include)

```
❯ http get :9191/metrics | grep temp
♯ HELP metabase_notification_temp_storage_total Number and type of temporary storage uses
♯ TYPE metabase_notification_temp_storage_total counter
metabase_notification_temp_storage_total{storage="memory",} 2.0
metabase_notification_temp_storage_total{storage="disk",} 2.0
metabase_notification_temp_storage_total{storage="truncated",} 2.0
```

dpsutton committed:
make predicates a bit more clojurey

dpsutton committed:
missed a refactor somehow

dpsutton committed:
better styling

thread the human readable size through

dpsutton committed:
styling changes

dpsutton committed:
bit of formatting

ranquild committed:
Merge remote-tracking branch 'origin/master' into size-insensitive-notification-temp-storage
